pkg/syncer/metadatasyncer.go

```diff
@@ -831,6 +839,318 @@ func InitMetadataSyncer(ctx context.Context, clusterFlavor cnstypes.CnsClusterFl
+				}
+			}
+		}
+	}
+	return spuWiseReservedMap, nil
+}
+
+// calculateVolumeSnapshotReservedForNamespace calculate the expected reserved value for volumesnapshot for namespace
+func calculateVolumeSnapshotReservedForNamespace(ctx context.Context, cnsOperatorClient client.Client,
+	namespace string) (map[string]*resource.Quantity, error) {
+	log := logger.GetLogger(ctx)
+	scTospuMap, err := fetchStorageClassToStoragePolicyMapping(ctx, cnsOperatorClient)
+	if err != nil {
+		return nil, err
+	}
+	log.Infof("Fetching VolumeSnapshots for namespace %q", namespace)
+	vsList := &snap.VolumeSnapshotList{}
+	err = cnsOperatorClient.List(ctx, vsList, &client.ListOptions{Namespace: namespace})
+	if err != nil {
+		return nil, err
+	}
+	pvcList := &v1.PersistentVolumeClaimList{}
+	volumeClaim := v1.PersistentVolumeClaim{}
+	volumeMap := make(map[string]v1.PersistentVolumeClaim)
+	spuWiseReservedMap := make(map[string]*resource.Quantity)
+	for _, vs := range vsList.Items {
+		if vs.DeletionTimestamp != nil {
+			log.Infof("volumesnapshot is marked for deletion, ignoring Name: %q, Namespace: %q",
+				vs.Name, vs.Namespace)
+			continue
+		}
+		if !*vs.Status.ReadyToUse {
+			log.Infof("Processing VolumeSnapshot Name: %q, Namespace: %q readyToUse %q",
+				vs.Name, vs.Namespace, vs.Status.ReadyToUse)
+			if len(pvcList.Items) == 0 {
+				log.Infof("Fetching PersistentVolumeClaim for Namespace: %q", vs.Namespace)
+				err := cnsOperatorClient.List(ctx, pvcList, &client.ListOptions{Namespace: namespace})
+				if err != nil {
+					return nil, err
+				}
+				for _, pvc := range pvcList.Items {
+					volumeMap[pvc.Name] = pvc
+				}
+			}
+			if vs.Spec.Source.PersistentVolumeClaimName != nil {
+				volumeClaim = volumeMap[*vs.Spec.Source.PersistentVolumeClaimName]
+				if spu, ok := scTospuMap[*volumeClaim.Spec.StorageClassName]; ok {
+					if spuWiseReservedMap[spu] == nil {
+						spuWiseReservedMap[spu] = resource.NewQuantity(0, resource.BinarySI)
+					}
+					spuWiseReservedMap[spu].Add(volumeClaim.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)])
+				}
+			}
+		}
+	}
+	return spuWiseReservedMap, nil
+}
+func updateStorageQuotaPeriodicSyncCR(ctx context.Context, cnsOperatorClient client.Client,
+	expectedReservedvalues []sqperiodicsyncv1alpha1.ExpectedReservedValues) error {
+	log := logger.GetLogger(ctx)
+	sqPeriodicSync := &sqperiodicsyncv1alpha1.StorageQuotaPeriodicSync{}
+	err := cnsOperatorClient.Get(ctx,
+		k8stypes.NamespacedName{
+			Name:      StorageQuotaPeriodicSyncInstanceName,
+			Namespace: common.GetCSINamespace()},
+		sqPeriodicSync)
+	if err != nil {
+		log.Errorf("failed to fetch StorageQuotaPeriodicSync instance with name %q Error: %+v",
+			StorageQuotaPeriodicSyncInstanceName, err)
+		return err
+	}
+	sqPeriodicSync.Status.ExpectedReservedValues = expectedReservedvalues
+	sqPeriodicSync.Status.LastSyncTimestamp = metav1.NewTime(time.Now())
```

Comment:
This timestamp should be time before fetching all resources, so please pass it from caller funtion